src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -49,79 +49,68 @@ Shard::Shard(DatabaseShard const& db, std::uint32_t index,
 }
 
 bool
-Shard::open(Section config, Scheduler& scheduler)
+Shard::open(Section config, Scheduler& scheduler, nudb::context& ctx)
 {
     assert(!backend_);
     using namespace boost::filesystem;
+    using namespace boost::beast::detail;
 
-    bool dirPreexist;
-    bool dirEmpty;
-    try
+    std::string const type (get<std::string>(config, "type", "nudb"));
+    auto factory {Manager::instance().find(type)};
+    if (!factory)
     {
-        if (!exists(dir_))
-        {
-            dirPreexist = false;
-            dirEmpty = true;
-        }
-        else if (is_directory(dir_))
-        {
-            dirPreexist = true;
-            dirEmpty = is_empty(dir_);
-        }
-        else
-        {
-             JLOG(j_.error()) <<
-                "path exists as file: " << dir_.string();
-            return false;
-        }
+        JLOG(j_.error()) <<
+            "shard " << index_ <<
+            ": failed to create shard store type " << type;
+        return false;
     }
-    catch (std::exception const& e)
+
+    boost::system::error_code ec;
+    auto const preexist {exists(dir_, ec)};
+    if (ec)
     {
         JLOG(j_.error()) <<
-            "shard " + std::to_string(index_) + " exception: " + e.what();
+            "shard " << index_ << ": " << ec.message();
         return false;
     }
 
+    config.set("path", dir_.string());
+    backend_ = factory->createInstance(
+        NodeObject::keyBytes, config, scheduler, ctx, j_);
+
     auto fail = [&](std::string msg)
     {
-        JLOG(j_.error()) <<
-            "shard " << std::to_string(index_) << " error: " << msg;
-
-        if (!dirPreexist)
-            removeAll(dir_, j_);
-        else if (dirEmpty)
+        if (!msg.empty())
```

Comment:
Don't you miss out on logging some errors this way? Line 106 will be empty string, which means nothing in the journal.

Reply:
No, the fail in `saveControl` already logged the error.